### PR TITLE
fix: added constraints for chart group name

### PR DESCRIPTION
**Pull request by Ashish-devtron**
# Description

Added check so that user cannot create chart group with mentioned condition.
1. Create chart group with name 'ch' or 'cha' or 'char', i.e. less than 5 characters.
2. Try to create one more chart group with same name that you used earlier.

User can't deploy multiple instances of charts with duplicate name.

Fixes https://github.com/devtron-labs/devtron/issues/1290
Fixes https://github.com/devtron-labs/devtron/issues/1289

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Tested locally with invalid constraints. Able to get the expected behaviour.

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -47,8 +47,8 @@
     "scripts": {
```
Comment (shivani170) [inline package.json:47]: PLease revert these changes

File: src/components/charts/modal/CreateChartGroup.tsx [hunk 2]
```diff
@@ -10,6 +10,7 @@ interface ChartGroupCreateState {
     name: { value: string; error: any[]; };
     description: string;
     loading: boolean;
+    result:ChartGroup[]
```
Comment (shivani170) [inline src/components/charts/modal/CreateChartGroup.tsx:13]: Please rename result to more specific name

File: src/components/charts/useChartGroup.ts [hunk 3]
```diff
@@ -228,22 +228,31 @@ export default function useChartGroup(chartGroupId = null): ChartGroupExports {
     
     async function validateData() {
         try {
+
             const nameRegexp = new RegExp(`^[a-z]+[a-z0-9\-\?]*[a-z0-9]+$`)
+            const allNames = state.charts.map((chart) => ( chart.name.value ))
```
Comment (shivani170) [inline src/components/charts/useChartGroup.ts:233]: Please add chart type

File: src/components/charts/useChartGroup.ts [hunk 4]
```diff
@@ -228,22 +228,31 @@ export default function useChartGroup(chartGroupId = null): ChartGroupExports {
...
+        let isNotValid  = allNames.filter((name,index)=>{
+             if(allNames.indexOf(name)!=index){
+                return index;
+             }
+             return false;
+        });
             let validated = true
             let tempCharts = state.charts.map((chart) => {
+                console.log(isNotValid.indexOf(chart.name.value));
```
Comment (shivani170) [inline src/components/charts/useChartGroup.ts:242]: Remove console

File: src/components/charts/useChartGroup.ts [hunk 5]
```diff
@@ -228,22 +228,31 @@ export default function useChartGroup(chartGroupId = null): ChartGroupExports {
...
+                if (!nameRegexp.test(chart.name.value) || !chart?.environment?.id ||isNotValid.length>0 )  {
                     validated = false
                 }
                 return {
                     ...chart,
                     name: {
                         value: chart.name.value,
                         error: nameRegexp.test(chart.name.value)
-                            ? ''
+                            ? (isNotValid.indexOf(chart.name.value)==-1?'':'Duplicate names found')
```
Comment (shivani170) [inline src/components/charts/useChartGroup.ts:255]: Can right (isNotValid.indexOf(chart.name.value) === -1 && 'Duplicate names found')

File: src/components/charts/useChartGroup.ts [hunk 6]
```diff
@@ -228,22 +228,31 @@ export default function useChartGroup(chartGroupId = null): ChartGroupExports {
...
+        });
             let validated = true
             let tempCharts = state.charts.map((chart) => {
+                console.log(isNotValid.indexOf(chart.name.value));
                 if (!chart.isEnabled) {
                     // dont consider disabled charts
                     return chart
                 }
-                if (!nameRegexp.test(chart.name.value) || !chart?.environment?.id) {
+                if (!nameRegexp.test(chart.name.value) || !chart?.environment?.id ||isNotValid.length>0 )  {
```
Comment (shivani170) [inline src/components/charts/useChartGroup.ts:247]: Please add formatting

File: src/components/charts/modal/CreateChartGroup.tsx [hunk 7]
```diff
@@ -81,6 +83,26 @@ export default class CreateChartGroup extends Component<CreateChartGroupProps, C
...
+        }
+
+        let isNameUsed = this.state.result.some((r)=>r.name===this.state.name.value);
+        if(isNameUsed){
+            showError({
+                message:'A chart group with this name already exists!'
+            });
+            return;
+        }
+
```
Comment (shivani170) [inline src/components/charts/modal/CreateChartGroup.tsx:101]: Please remove all unnecessary lines

File: src/components/charts/modal/CreateChartGroup.tsx [hunk 8]
```diff
@@ -81,6 +83,21 @@ export default class CreateChartGroup extends Component<CreateChartGroupProps, C
             })
         }
 
+        if(this.state.name.value.length<5){
+            showError({
```
Comment (shivani170) [inline src/components/charts/modal/CreateChartGroup.tsx:87]: This is not the right way, Instead of this try to set the state for `name: { value: "", error: [] }` and update the error

File: src/components/charts/modal/CreateChartGroup.tsx [hunk 9]
```diff
@@ -1,55 +1,54 @@
...
 
-            if (!startAndEndAlphanumericRegex.test(event.target.value) && !(event.target.value.length ==1)) {
+            if (!startAndEndAlphanumericRegex.test(event.target.value) && !(event.target.value.length == 1)) {
                 errors.push('Start and end with an alphanumeric character only')
             }
 
-            if (event.target.value.trim().indexOf(" ") >= 0) {
-                errors.push('Do not use \'spaces\'')
+            if (event.target.value.trim().indexOf(' ') >= 0) {
+                errors.push("Do not use 'spaces'")
```
Comment (shivani170) [inline src/components/charts/modal/CreateChartGroup.tsx:51]: Keep all the messages in constants

File: src/components/charts/modal/CreateChartGroup.tsx [hunk 10]
```diff
@@ -60,91 +59,136 @@ export default class CreateChartGroup extends Component<CreateChartGroupProps, C
...
                 name: {
                     ...this.state.name,
-                    error: ['This is a required field']
-                }
+                    error: ['This is a required field'],
+                },
             })
         }
 
+        if (this.state.name.value.length < 5) {
```
Comment (shivani170) [inline src/components/charts/modal/CreateChartGroup.tsx:82]: extract out these in a separate function and handle isValid cases.

File: src/components/charts/modal/CreateChartGroup.tsx [hunk 11]
```diff
@@ -60,91 +59,136 @@ export default class CreateChartGroup extends Component<CreateChartGroupProps, C
...
+
+                <label className="form__row">
+                    <span className="form__label">Description</span>
+                    <textarea
+                        className="form__input form__input--textarea"
+                        name="description"
+                        value={this.state.description}
+                        placeholder="Enter a short description for this group."
+                        autoFocus={true}
+                        tabIndex={1}
```
Comment (shivani170) [inline src/components/charts/modal/CreateChartGroup.tsx:178]: Its tabIndex should not be 1

File: src/components/charts/useChartGroup.ts [hunk 12]
```diff
@@ -225,25 +236,35 @@ export default function useChartGroup(chartGroupId = null): ChartGroupExports {
...
+                if (!nameRegexp.test(chart.name.value) || !chart?.environment?.id || isNotValid?.length > 0) {
                     validated = false
                 }
                 return {
                     ...chart,
                     name: {
                         value: chart.name.value,
                         error: nameRegexp.test(chart.name.value)
-                            ? ''
+                            ? isNotValid.indexOf(chart.name.value) === -1
```
Comment (shivani170) [inline src/components/charts/useChartGroup.ts:265]: Avoid ternary nesting instaed extract out in a separate function for more rediability

File: src/components/charts/modal/CreateChartGroup.tsx [hunk 13]
```diff
@@ -60,91 +59,136 @@ export default class CreateChartGroup extends Component<CreateChartGroupProps, C
...
+                        onChange={this.handleDescriptionChange}
+                        required
+                    />
+                    <span className="form__error">
+                        {/* {showError && !this.state.isValid.appName
```
Comment (shivani170) [inline src/components/charts/modal/CreateChartGroup.tsx:183]: Remove commented unused code

File: src/components/charts/modal/CreateChartGroup.tsx [hunk 14]
```diff
@@ -60,91 +59,141 @@ export default class CreateChartGroup extends Component<CreateChartGroupProps, C
...
+
+    checkValid() {
+        if (this.state.name.value.length < 5) {
+            return false
+        }
+
+        let isNameUsed = this.state.charts.some((r) => r.name === this.state.name.value)
+        if (isNameUsed) {
+            showError({
+                message: 'A chart group with name ' + this.state.name.value + ' already exists!',
```
Comment (shivani170) [inline src/components/charts/modal/CreateChartGroup.tsx:80]: Can use template literals here instead of concatenation

File: src/components/charts/modal/CreateChartGroup.tsx [hunk 15]
```diff
@@ -60,91 +59,141 @@ export default class CreateChartGroup extends Component<CreateChartGroupProps, C
...
+            })
     }
 
     //TODO: setting state from props is anti-pattern. what is the need of name and description in if condition?
     componentDidMount() {
         if (this.props.chartGroupId && this.props.name) {
-            this.setState({ name: { value: this.props.name, error: [] }, description: this.props.description || "" })
+            this.setState({ name: { value: this.props.name, error: [] }, description: this.props.description || '' })
         }
+        getChartGroups().then((res) => this.setState({ charts: res.result.groups }))
```
Comment (shivani170) [inline src/components/charts/modal/CreateChartGroup.tsx:140]: Extract this out in a separate function and use async await

File: src/components/charts/modal/CreateChartGroup.tsx [hunk 16]
```diff
@@ -60,91 +59,141 @@ export default class CreateChartGroup extends Component<CreateChartGroupProps, C
...
+                        name="name"
+                        value={this.state.name.value}
+                        placeholder="e.g. elastic-stack"
+                        autoFocus={true}
+                        tabIndex={1}
+                        onChange={this.handleNameChange}
+                        required
+                    />
+                    <span className="form__error">
+                        {this.state.name.error.map((itm) => {
```
Comment (shivani170) [inline src/components/charts/modal/CreateChartGroup.tsx:168]: rename itm to appropriate name

File: src/components/charts/useChartGroup.ts [hunk 17]
```diff
@@ -225,25 +236,38 @@ export default function useChartGroup(chartGroupId = null): ChartGroupExports {
             showError(err)
         }
     }
-    
+
+    function isValidName(value: string, invalidNames: string[]): string {
+        let err = invalidNames.indexOf(value) === -1 ? '' : 'Duplicate names found'
```
Comment (shivani170) [inline src/components/charts/useChartGroup.ts:241]: invalidNames.indexOf(value) === -1 ? '' : 'Duplicate names to `invalidNames.indexOf(value) === -1 && 'Duplicate names found`

File: src/components/charts/useChartGroup.ts [hunk 18]
```diff
@@ -225,25 +236,38 @@ export default function useChartGroup(chartGroupId = null): ChartGroupExports {
             showError(err)
         }
     }
-    
+
+    function isValidName(value: string, invalidNames: string[]): string {
+        let err = invalidNames.indexOf(value) !== -1 &&'Duplicate names found'
```
Comment (milindk8) [inline src/components/charts/useChartGroup.ts:241]: add space & move to const

File: src/components/charts/useChartGroup.ts [hunk 19]
```diff
@@ -225,25 +236,38 @@ export default function useChartGroup(chartGroupId = null): ChartGroupExports {
...
+        return err
+    }
+
     async function validateData() {
         try {
             const nameRegexp = new RegExp(`^[a-z]+[a-z0-9\-\?]*[a-z0-9]+$`)
+
+            const allNames = state.charts.map((chart) => chart.name.value)
+            let invalidNames = allNames.filter((name, index) => {
```
Comment (milindk8) [inline src/components/charts/useChartGroup.ts:250]: const

**Review by milindk8 — APPROVED** [review]

**Review by shivani170 — APPROVED** [review]

**Review by shivani170 — APPROVED** [review]

**Review by sohel-rp — APPROVED** [review]
